spark/src/main/scala/org/apache/spark/sql/delta/schema/ImplicitMetadataOperation.scala

```diff
@@ -206,25 +206,68 @@ object ImplicitMetadataOperation {
    * @return Merged schema
    */
   private[delta] def mergeSchema(
+      spark: SparkSession,
       txn: OptimisticTransaction,
       dataSchema: StructType,
       isOverwriteMode: Boolean,
       canOverwriteSchema: Boolean): StructType = {
     if (isOverwriteMode && canOverwriteSchema) {
       dataSchema
     } else {
-      val fixedTypeColumns =
-        if (GeneratedColumn.satisfyGeneratedColumnProtocol(txn.protocol)) {
-          txn.metadata.fixedTypeColumns
-        } else {
-          Set.empty[String]
-        }
+      checkDependentExpressions(spark, txn.protocol, txn.metadata, dataSchema)
+
       SchemaMergingUtils.mergeSchemas(
         txn.metadata.schema,
         dataSchema,
-        fixedTypeColumns = fixedTypeColumns,
         allowTypeWidening = TypeWidening.isEnabled(txn.protocol, txn.metadata))
     }
   }
 
+  /**
+   * Finds all fields that change between the current schema and the new data schema and fail if any
+   * of them are referenced by check constraints or generated columns.
+   */
+  private def checkDependentExpressions(
+      sparkSession: SparkSession,
+      protocol: Protocol,
+      metadata: actions.Metadata,
+      dataSchema: StructType): Unit =
+  SchemaMergingUtils.transformColumns(metadata.schema, dataSchema) {
+    case (fieldPath, currentField, Some(updateField), _)
+      // This condition is actually too strict, structs may be identified as changing because one
+      // of their field is changing even though that field isn't referenced by any constraint or
+      // generated column. This is intentional to keep the check simple and robust, esp. since it
+      // aligns with the historical behavior of this check.
+      if !SchemaMergingUtils.equalsIgnoreCaseAndCompatibleNullability(
+        currentField.dataType,
+        updateField.dataType
+      ) =>
+        val columnPath = fieldPath :+ currentField.name
+        // check if the field to change is referenced by check constraints
+        val dependentConstraints =
+          Constraints.findDependentConstraints(sparkSession, columnPath, metadata)
```

Comment:
This seems a bit inefficient. Won't this result in the constraints and generated columns being parsed over and over again? Probably not a big issue in practice though.

Reply:
True but it allows unifying the checks we do here during schema evolution and the checks in ALTER TABLE so that both use the same helpers.

We can improve in the future, in this change I'd rather focus on correctness and reuse helpers that have been field-tested already.